forms/validators/answer.py

```diff
@@ -28,17 +30,37 @@ def __call__(self, value):
 
 class RequiredFormFieldValidator:
     def __call__(self, value):
-        for section in value["form"].sections.all():
-            for field in section.fields.all():
-                if not field.required:
-                    continue
-                found = False
-                if (
-                    field.identifier
-                    in list(value["entries"][section.identifier].keys())
-                    and value["entries"][section.identifier][field.identifier] != ""
-                ):
-                    found = True
-                    continue
-                if not found:
+        self.required_validator(
+            value["entries"],
+            Field.objects.filter(section__form=value["form"], required=True),
+        )
+
+    def required_validator(
+        self, entries, required_fields, section_identifier=None, field_identifier=None
+    ):
+        if "sections" in entries:
+            self.required_validator(
+                entries["sections"], required_fields, None, field_identifier
+            )
+        if "fields" in entries:
+            self.required_validator(
+                entries["fields"], required_fields, section_identifier, field_identifier
+            )
+        if section_identifier is not None:
+            for entry in entries:
+                if required_fields.filter(
+                    section__identifier=section_identifier, identifier=entry
+                ) and (entries[entry] == ""):
                     raise ValidationError(code="required")
+        if isinstance(entries, list):
+            for i, entry in enumerate(entries):
+                return self.required_validator(
+                    entry, required_fields, section_identifier=section_identifier
+                )
+        if isinstance(entries, str):
```

Comment:
I'm getting a 500 with "argument of type 'bool' is not iterable" on line 41 when submitting a new application with no required fields. I assume it's because this clause is somehow in charge of catching all valid values, but strings aren't the only allowed types at the moment. (The primary exception to this currently are checkbox fields - single checkbox fields output either true or false, multiple checkbox fields an array of IDs.)

I think that this function should focus on only parsing the section structure. Identifying what is a field and what isn't is easy – the only location fields can ever be in in a valid submission from the UI is the "fields" key, containing a one-dimensional simple array of "value + extra value" simple objects. Thus, when those fields are found, line 46 can call the "validate this field" function with something similar to lines 50-54 moved to it, and the rest of the function can just deal with the sections.

Reply:
The error hasn't changed since the last fix – still getting a 500 when a boolean value is encountered.

Reply:
Now this should be fixed

Reply:
Good news: indeed, the save now goes through.
Bad news: it didn't seem to catch an empty value in a required field 😅 I set the applicant type field as required (while removing the frontend check for it) but the save still went through with nothing selected (value "" in path `entries.sections["hakijan-tiedot"][0].fields.hakija.value`).

Reply:
Also, I took another look at the current code and I'm not sure if this would still handle one of those multi-selection fields properly. I'll test it later myself when the basic case is working, but you can give it a go with the value `[]` too. (In case of an array, "required" should mean at least one value is given.)

Reply:
Was it empty field or ""

Reply:
Just an empty string.

Reply:
I tried same kind of dict where is list without any probles

Reply:
No change with the newest iteration, the save still goes through.

Here's the current form structure (as given by the plot search list GET) and the empty application I'm sending into the answers POST endpoint in case there's something unexpected in the way they're structured. The only required field for now is the one I mentioned above, for which I'm not sending a value for at all.
https://gist.github.com/EmiliaMakelaVincit/5ab8a3ea61e34beb0a3c58e2bf62ff1e

Reply:
Alright, with this latest version that case works properly now as well! I'd say that should make this mergeable.